Internal database reimplementation

alexrutar opened the pull request:
This PR re-implements a lot of the database internals. Most notably, canonical identifiers are no longer unique elements of the database: instead, each canonical identifier corresponds to a _tree_ of rows in the database, and each row contains the row-ids of the parents and children (if any).

The entries in the `CitationKeys` index still point to unique row-ids. There is a new invariant: only one row can be active at a given moment (i.e., all elements in `CitationKeys` pointing to a given canonical id must also point to the same row).

The reason for this change is to implement support for undo/redo; i.e. #233 Now, undo and redo are first-class and simply amount to changing the row ids in the `CitationKeys` table to point to a different entry in the same tree. Undo always works (unless you are at the root) and redo will require a prompt if there are divergent changes.

So far, the following additional changes/simplifications have been made.

1. The implementation of `autobib delete` is a lot simpler. We just update a given node to replace it with a 'deletion' marker. This is "soft deletion". We no longer delete the elements in `CitationKeys`. If an attempt is made to retrieve a soft-deleted entry, the appropriate methods will report an error. There is also `autobib delete --hard`, which evicts all mention of a given key, deletes aliases, etc. Also, we no longer worry about deduplicating the inputs (this is what is expected, e.g. with `rm`: the first deletion will succeed, and then subsequent ones will return errors).
2. `autobib merge` has been removed. Instead, there is `autobib delete --replace <key>`, which soft-deletes a record and indicates that it should be replaced by another record. Attempting to retrieve a soft-deleted key in this way will return an error message which suggests a replacement key.
3. `autobib update` has been changed so that the update can be performed in one of three ways: from the remote data, from a bibtex file using `--from-bibtex`, or from another reford using `--from-key`. The `--from-bibtex` argument was previously called `--from`. I also fixed it so that data provided by a from a bibtex file or from another record cannot be used to update an entry in the `NullRecords` table.
4. `autobib undo` and `autobib redo` have been implemented. Since every state has a unique parent, `autobib undo` requires no new information to succeed. In contrast, `autobib redo` requires extra information since each state can have multiple children. Therefore, `autobib redo` accepts an optional `INDEX` argument, which provides the index of the divergent change. If `INDEX` is not provided and there is exactly one child, we can redo to that child anyway; otherwise, `INDEX` is required.
5. A new command `autobib log` has been added. By default, it only shows the immediate history of the active version, but it also accepts a `--all` argument, which shows the full history of a given identifier, including divergent changes.
6. I added a missing `NOT NULL` constraint in the `CitationKeys` table.

I will update this list as this PR continues.

TODOs:

- [x] Implement database migration
  - [x] The 'Records' table migration is quite simple, since the new table has default values for the new columns.
  - [x] Unwind previously merged providers (i.e. canonical ids in `CitationKeys` which point to a different canonical id in `Records`), and replace them with entries equivalent to `autobib delete --replace`.
  - [x] Rename the `CitationKeys` table to the new name `Identifiers`; there is a new NOT NULL constraint so it requires a bulk re-insert.
  - [x] Delete the `Changelog` table, which is no longer used at all.
- [x] Update internal documentation, especially adding docs about the new state representation
  - [x] Check for naming consistency in help messages
  - [x] Add `info!` and `debug!` logging where appropriate
- [x] Implement #252; this should be a two step (1) compute provenance and (2) essentially run `autobib delete --replace`
- [x] Fix the implementation of `autobib util check`
- [x] Update changelog
- [x] Add tests for the new behaviour
- [x] Change `autobib delete`
- [x] Remove `autobib merge`
- [x] Change `autobib update`
  - [x] Apply normalizations on incoming remote data for `autobib update`
- [x] Implement `autobib undo` and `autobib redo`
- [x] Implement `autobib reset`
  - [x] Reset to a specific revision
  - [x] Reset to a specific timestamp
- [x] Better logging for `undo` and `redo` and `reset`, printing information about the new state after the undo
- [x] Correctly handle `autobib update` and `autobib edit` when targeted towards a 'deletion' record
  - [x] This adds a `autobib update --revive` option to update even if the record does not exist. I've also added a `autobib hist revive` command as a general command for adding new versions of existing records. The difference between `autobib update --revive` is that it retrieves data from remote, whereas `autobib hist revive` asks you to input the data manually. I think `autobib edit` should just fail if the data does not exist.
- [x] Reconsider `autobib:local` automatic identifiers
  - [x] While cleaning up the code, I have temporarily removed the automatic alias creation associated with local identifiers. Is this reasonable?
- [x] Changes to `autobib edit`
  - [x] If a normalization is specified, no longer open the editor.
  - [x] Add headless edit methods.
- [x] Correctly handle files, if files exist on soft deletion. Perhaps `autobib find --attachments` should also include soft-deleted entries? Maybe warn if files exist, and say the user should deal with them? Differences between soft / hard deletion? Since merging no longer changes the canonical identifier, we no longer need to move files around. (probably defer this to a new PR)
- [x] Implement #249 (probably defer this to a new PR)
- [x] Implement #251 (probably defer this to a new PR)
- [x] Add ways to prune old entries (defer?)
  - [x] Delete non-active versions which are older than a given age
  - [x] Only keep `n` most recent changes (plus their children)
  - [x] Prune void entries (deleting if present)
- [x] Add new alias subcommand `autobib alias update`
  - [x] Added `--update-aliases` option which deletes/renames the aliases depending on whether `--replace` is specified
- [x] Add 'global' changelog operations
  - [x] Showing all recent changes, up to some limit (count, after time, ...)
  - [x] Bulk options for `autobib reset`, for example to undo bad imports.
- [x] Add a new command `autobib log`
  - [x] Provide option show the history / sequence of parents of the current node.
  - [x] Options: `--all`: traverse deletion boundaries; `--tree` show full tree vs only history; `--oneline` use a short diff option for each branch and print one line per branch. Maybe the printed info should be customizable with a template? Need to be careful with multi-line printing.
  - [x] Implement `--oneline` mode. This requires some sort of diffing mechanism, r.e. #212 (defer)
  - [x] Improve formatting style / appearance. Make sure to only write styles if printing to an interactive terminal
  - [x] A '--json' mode, which outputs a JSON representation of the full history, starting at the root, and then with nested entries for the children (or as a list, depending on whether `--all` is passed or not) (defer)
- [x] Allow accessing / reading past states by key (checking if valid, etc.), perhaps as hexadecimal
  - [x] Allow changing to a specific hexadecimal target with `autobib hist reset <key> <hex-target>`
  - [x] Use these states in the output of `autobib log`

alexrutar commented:
Nice looks like GitHub programming broke the macos runner.

https://github.com/actions/runner/issues/4134

https://github.com/orgs/community/discussions/180160

https://github.com/actions/runner-images/issues/13341

etc.

wupr commented:
Regarding a new PR -- one possibility is to start a temporary development branch and create PRs to that branch first, which would allow the flexibility of having incomplete code/docs and even doing some force-pushes.

Removing automatic alias creation sounds fair.

alexrutar commented:
Also see:

- https://github.com/autobib/autobib/blob/db-rework/ARCHITECTURE.md
- https://github.com/autobib/autobib/blob/db-rework/docs/data_model.md

wupr commented on docs/data_model.md:
```diff
@@ -0,0 +1,113 @@
+   For example, this will cause `autobib get` to automatically retrieve new data for the row.
+   It is uncommon for a record to be in this state, but this can be attained manually with `autobib hist void`, or is created automatically with `autobib hist rewind-all` or `autobib hist reset --before` when the threshold time precedes the existence of the record in the database.
+
+Note that `void` entries still contain some state: it preserve aliases, and still tracks the *prov-id* which can be used to more efficiently look-up new data.
```
"preserve" --> "preserves"

wupr commented on docs/data_model.md:
```diff
@@ -0,0 +1,113 @@
+
+The simplest operations regarding revisions are `autobib hist undo` and `autobib hist redo`.
+Undo sets the active state to the parent state, if the parent state exists.
+Redo sets the active state to the child if there is exactly one child; otherwise, it is required to pass an explicit index indicating which child to follow.
```
Is this still true?

wupr commented on docs/data_model.md:
```diff
@@ -0,0 +1,113 @@
+Redo sets the active state to the child if there is exactly one child; otherwise, it is required to pass an explicit index indicating which child to follow.
+
+For more complex operations involving the edit-tree, the output of `autobib log --tree` can be useful.
+This prints a branch diagram showing the relationship between all of the states ordered in reverse chronological order.
```
"ordered in reverse chronological order" --> "in reverse chronological order"

wupr commented on src/app/cli.rs:
```diff
@@ -477,22 +558,152 @@ impl Command {
+    /// modification greater than the provided time.
+    ///
+    /// Use caution! The modification time may not correspond to the database state at the provided
+    /// date-time if you have used `autobib undo/redo/reset`, since these methods only change the
```
`autobib hist (undo|redo|reset)`

wupr commented on src/app/cli.rs:
```diff
@@ -477,22 +558,152 @@ impl Command {
+    },
+    /// Move the database back in time.
+    ///
+    /// This is the same as calling `autobib reset --before` on every active entry in the database with
```
`autobib hist reset --before`

wupr commented on src/app/cli.rs:
```diff
@@ -164,57 +178,82 @@ pub enum Command {
+    /// By default, this performs a soft delete, where the current data and keys are retained
+    /// but future attempts to read them will result in an error. The data can be recovered with
+    /// `autobib hist undo`. The key provided with the `--replace` option will be used to suggest
+    /// replacements.
```
```suggestion
    /// replacements when the deleted data is requested.
```

wupr commented on src/app/cli.rs:
```diff
@@ -164,57 +178,82 @@ pub enum Command {
+        /// Hard deletion, which removes all history and aliases, and cannot be undone.
+        #[arg(long, group = "delete_mode")]
+        hard: bool,
+        /// Update aliases, either deleting or changing them to point to the new row if `--replace` is specified.
```
```suggestion
        /// Update aliases, making them point to the new record if `--replace` is specified, or deleting them otherwise.
```

wupr commented on docs/changelog/next.md:
```diff
@@ -1 +1,49 @@
+  Passing redundant identifiers will now result in an error.
+- `autobib merge` has been removed.
+  It is no longer possible to merge records together.
+  This functionality has been replaced by `autobib delete --replace <replacement>`, which performs a soft delete and stores a replacement identifier.
```
I don't think `autobib delete --replace <replacement>` completely replaces `autobib merge`, as the user can't easily merge even non-conflicting fields now. We should explain this.

alexrutar commented:
In the new version, merging is no longer possible - precisely for the reason that you have identified that when you merge, you lose information about aliases (so "unmerging" is impossible without some sort of procedure to decide where to point the aliases).

I was hoping that `autobib delete --replace` would be a sufficiently ergonomic fallback. Now instead of merging, you should run `autobib delete --replace` with the thing you would have merged into. Then, if you run `autobib get` with the "merged" record, in the current implementation it will fail but provide a warning indicating that it was merged into another record, and tell you what to use instead.

This behaviour could be updated to "auto-follow the replacement but warn"? Then `autobib merge` could be re-introduced, essentially as a wrapper around
```
autobib update <keep> --from-record <delete>
autobib delete <delete> --replace <keep>
```

The reason why I did not implement auto-follow was: (1) what about chain replacement (a -> b -> c) which makes following recursive; and (2) it seems weird to auto-follow if the target might not be in the database.

(Edit: maybe (2) could be implemented by having a special 'merged into' column, with a foreign key relation on the `name` column of the `Identifiers` table.)

I originally thought it would be nice for the replacement to point to a literal row in the 'Records' table - but there are foreign key issues here - and also what to do when updating rows?

Keen to hear your thoughts here since you seem to be using `merge` more than me!

(Edit 2: another funny foreign key trick which might work: point to the `record_key` column in 'Identifiers' and use `ON UPDATE CASCADE ON DELETE RESTRICT / SET NULL`; because 'Identifiers' is always updated this will correctly cascade the update.)

wupr commented on src/app/cli.rs:
```diff
@@ -227,77 +241,144 @@ pub enum Command {
+        identifier: RecordId,
+        /// Replace with another identifier.
+        #[arg(short, long, group = "replace_target")]
+        with: Option<RecordId>,
```
We could Clap attribute `value_name = "IDENTIFIER"` for `identifier` and `with`

wupr reviewed (approved):
@alexrutar Happy new year! And very sorry about the long delay. I was taking the time to re-learn the code base but then got distracted by other things. I think I'll just put together some comments now instead of blocking this any further. Most functionalities I tried appeared to be working as expected. (The only thing I haven't tested is `autobib import`.)

#### Housekeeping

- `src/db/migrate/v1/records_new.sql` seems to be redundant (replaced by `src/db/migrate/v1/create_table_tmp_records.sql`).
- `tests/resources/dedup` needs to be renamed now since we no longer have `autobib dedup`.

#### Documentation

- Should we bring more attention to the change of terminology in the changelog?
- We are changing a lot of CLI options. We could mention these at one stroke and refer the user to the help messages, even if we don't enumerate all of them.
- In view of the upcoming docs website, it might make thing tidier if we moved `ARCHITECTURE.md` to `docs/architecture.md` and renamed `docs/data_model.md` to `docs/data-model.md`?

#### CLI

- How about making `autobib edit --touch abc` --> `autobib hist touch abc`, and `autobib hist touch-all` --> `autobib hist touch --all`?
- Similarly, how about `autobib hist reset --before time abc` --> `autobib hist rewind abc time`, and `autobib hist rewind-all` --> `autobib hist rewind --all`?
- ~~`autobib replace`: I thought the "target identifier" was the "current" one, and the "original identifier" was the "incoming" one, until I tried the command and noticed it was implemented the other way. And I can find arguments for both options. If the target is being remotely retrieved, then it's intuitively the incoming one. But it is the target record's history that has data written to it, so it seems like the target should be the current one.~~ I think you've changed this since I last read the code.

I have more notes, but they're less important and can be separate issues or PRs after this is merged.

alexrutar commented on src/app/cli.rs:
```diff
@@ -169,52 +175,60 @@ pub enum Command {
         /// Set "eprint" and "eprinttype" BibTeX fields from provided fields.
-        #[arg(long, value_delimiter = ',', value_name = "FIELD_NAME")]
+        #[arg(long, value_delimiter = ',', value_name = "FIELD_KEY")]
         set_eprint: Vec<String>,
```
I've inlined these comments now instead of putting them at the top-level documentation! I think this is generally speaking a bit easier to read.

alexrutar commented on src/app/cli.rs:
```diff
@@ -336,16 +414,16 @@ pub enum Command {
         #[arg(long, group = "output")]
         retrieve_only: bool,
-        /// Only print the citation keys which were found (sorted and deduplicated).
+        /// Only print the identifiers keys which were found (sorted and deduplicated).
```
In this specific case yes, since they're actual citation keys found in the file.

alexrutar commented on src/record/key.rs:
```diff
@@ -340,15 +351,22 @@ impl RemoteId {
         Self::new_unchecked(full_id, PROVIDER_LEN)
     }
+
+    pub fn get_ref(&self) -> RemoteId<&str> {
```
Hmm, I think `as_deref` isn't terrible in this case, except for the fact that there are no `T: Deref` bounds around currently. Perhaps I will rename to that for now (and it can be renamed later again if a better name shows up.)

alexrutar commented on src/app/cli.rs:
```diff
@@ -227,99 +241,166 @@ pub enum Command {
     /// Import records from a BibTeX file.
+    ///
+    /// The implementation automatically determines a remote identifier from the data, using
+    /// your `preferred_providers` config setting and with unspecified fallback if there is no
```
I've fixed the phrasing!

alexrutar commented:
> @alexrutar Happy new year! And very sorry about the long delay. I was taking the time to re-learn the code base but then got distracted by other things. I think I'll just put together some comments now instead of blocking this any further. Most functionalities I tried appeared to be working as expected. (The only thing I haven't tested is `autobib import`.)
> 
> #### Housekeeping
> * `src/db/migrate/v1/records_new.sql` seems to be redundant (replaced by `src/db/migrate/v1/create_table_tmp_records.sql`).
> * `tests/resources/dedup` needs to be renamed now since we no longer have `autobib dedup`.
> 
> #### Documentation
> * Should we bring more attention to the change of terminology in the changelog?
> * We are changing a lot of CLI options. We could mention these at one stroke and refer the user to the help messages, even if we don't enumerate all of them.
> * In view of the upcoming docs website, it might make thing tidier if we moved `ARCHITECTURE.md` to `docs/architecture.md` and renamed `docs/data_model.md` to `docs/data-model.md`?
> 
> #### CLI
> * How about making `autobib edit --touch abc` --> `autobib hist touch abc`, and `autobib hist touch-all` --> `autobib hist touch --all`?
> * Similarly, how about `autobib hist reset --before time abc` --> `autobib hist rewind abc time`, and `autobib hist rewind-all` --> `autobib hist rewind --all`?
> * ~`autobib replace`: I thought the "target identifier" was the "current" one, and the "original identifier" was the "incoming" one, until I tried the command and noticed it was implemented the other way. And I can find arguments for both options. If the target is being remotely retrieved, then it's intuitively the incoming one. But it is the target record's history that has data written to it, so it seems like the target should be the current one.~ I think you've changed this since I last read the code.
> 
> I have more notes, but they're less important and can be separate issues or PRs after this is merged.

I've implemented this as `autobib hist touch --id abc` or `autobib hist rewind --id abc`! I think clap works a bit more cleanly with this type of syntax - plus it leaves flexibility in the future to allow multiple / combined targets (like, `--id abc --matches zbl:*` or similar).
